src/k2/transport/Payload.cpp

```diff
@@ -36,7 +36,7 @@ Payload::PayloadPosition::PayloadPosition(_Size bIdx, _Size bOff, size_t offset)
     // Empty payload will have position of (bidx=0, boff=0, off=0)
 }
 
-Payload::Payload(BinaryAllocatorFunctor allocator):
+Payload::Payload(seastar::lw_shared_ptr<BinaryAllocator> allocator):
```

Comment:
I think you should use allocator objects instead of shared pointer. The shared_ptr approach is more expensive since it requires allocation

Reply:
The purpose of a shared pointer here is to allow multiple payloads to use the same allocator to avoid allocating a BinaryAllocator for each individual payload. It is more expensive to use an allocator object for each payload in such a use case.

Reply:
I think the shared_ptr is more expensive. A shared pointer copies a pointer and increments a shared counter. The functor object on the other hand is just a copy of a pointer (the function underneath).

Moreover, the shared pointer does require allocation, as you do in the rest of the code base (calling make_lw_shared). You could "share" the shared_ptr, but this is just as expensive since just copying the object since it is so light.
Lastly, the shared_ptr approach also now requires 2x dereferencing (1 for the shared_ptr, 2 for the function in the object) whereas with the object it is just 1.

Reply:
I see, this is because the BinaryAllocator object is so light that the direct object copy here is less expensive than copying a shared pointer and updating the object reference count in the shared pointer.

Will change to use the object directly.

Reply:
done